R/settings.R

```diff
@@ -77,6 +78,72 @@ read_config_file <- function(config_file) {
   config
 }
 
+validate_config_file <- function(config, config_file, defaults) {
+  matched <- names(config) %in% names(defaults)
+  if (!all(matched)) {
+    warning("Found non-setting objects in config '", config_file, "': ", toString(names(config)[!matched]))
+  }
+
+  check_types <- intersect(names(config), names(defaults))
```

Comment:
Wouldn't this test throw on character exclusions?

It seems more expressive to explicitly list the keys and what they should be though. That would also allow for regex validation for the nolint configs.

Reply:
It increases the complexity a lot for not much benefit IMO.

Reply:
I'd argue it makes the actual validations much more clear. Currently, you have to go look up the definition of `default_settings` to figure out what types are expected.

NB regex validation could be very simple to implement. Not sure how nice the error messages are if you provide an invalid regex to `exclude` and friends

Reply:
Okay, fair enough.
In that case, I would disable the check for exclusions because it can vary in types. I see no need to introduce a breaking change.

Reply:
> I would disable the check for exclusions because it can vary in types

Yes, the current state is flexible enough -- tests pass without needing to adapt any old tests.

Currently, I only added tests of errors -- do you think it's worth adding tests of valid configs in this PR as well?

Reply:
Are they? What if `exclusions` is given as a character?

Reply:
Added some explicit tests for clarity, PTAL